modal/image.py

```diff
@@ -650,6 +650,27 @@ def my_build_function():
         """
         return self.extend(build_function=(raw_function, kwargs))
 
+    def env(self, vars: dict[str, str]) -> "_Image":
+        """Sets the environmental variables of the image.
+
+        **Example**
+
+        ```python notest
+        image = (
+            modal.Image.conda()
+                .env(CONDA_OVERRIDE_CUDA="11.2")
+                .conda_install(["jax", "cuda-nvcc"], channels=["conda-forge", "nvidia"])
+                .pip_install(["dm-haiku", "optax"])
+        )
+        ```
+        """
+        # TODO environment replacement, '$' is disallowed for now
+        if any("$" in val for (key, val) in vars.items()):
+            raise ValueError("'$' not allowed in values")
```

Comment:
This would be better detected on the worker-side rather than the client-side, since there are other lower-level image builder functions like `from_dockerfile()` that check for it. If you raise an error in the builder, it should show up in the client.

Also this error message might be improved: rather than just saying `'$' not allowed in values`, you could give more context, like `environment variable substitution is not yet implemented in definition of "{VAR_NAME_HERE}"`

Reply:
The client currently puts the values in shlex.quote() before sending it, is this an issue?

.env({"foo": "$bar"}) becomes "ENV foo='$bar'"

Reply:
Good point! I just tried it in Docker. It seems like the quotes themselves aren't a problem, but `shlex.quote()` won't do what you want with newlines.

```dockerfile
FROM alpine
ENV FOO="foo\n bar"
RUN echo $FOO
```

Then if I get a shell into the image and run it:

```
/ # echo $FOO
foo\n bar
```